quil-rs/src/instruction/qubit.rs

```diff
@@ -1,19 +1,122 @@
-use std::fmt;
+use std::sync::Arc;
 
-use serde::{Deserialize, Serialize};
+use crate::quil::{Quil, ToQuilError};
 
-#[derive(Clone, Debug, Eq, Hash, PartialEq, Serialize, Deserialize)]
+#[derive(Clone, Debug, Eq, Hash, PartialEq, PartialOrd, Ord)]
 pub enum Qubit {
     Fixed(u64),
+    Placeholder(QubitPlaceholder),
     Variable(String),
 }
 
-impl fmt::Display for Qubit {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+impl Qubit {
+    pub(crate) fn resolve_placeholder<R>(&mut self, resolver: R)
+    where
+        R: Fn(&QubitPlaceholder) -> Option<u64>,
+    {
+        if let Qubit::Placeholder(placeholder) = self {
+            if let Some(resolved) = resolver(placeholder) {
+                *self = Qubit::Fixed(resolved);
+            }
+        }
+    }
+}
+
+impl Quil for Qubit {
+    fn write(
+        &self,
+        writer: &mut impl std::fmt::Write,
+        fall_back_to_debug: bool,
+    ) -> std::result::Result<(), crate::quil::ToQuilError> {
         use Qubit::*;
         match self {
-            Fixed(value) => write!(f, "{value}"),
-            Variable(value) => write!(f, "{value}"),
+            Fixed(value) => write!(writer, "{value}").map_err(Into::into),
+            Placeholder(_) => {
+                if fall_back_to_debug {
+                    write!(writer, "{:?}", self).map_err(Into::into)
+                } else {
+                    Err(ToQuilError::UnresolvedQubitPlaceholder)
+                }
+            }
+            Variable(value) => write!(writer, "{value}").map_err(Into::into),
         }
     }
 }
+
+type QubitPlaceholderInner = Arc<()>;
```

Comment:
The comments on [this StackOverflow question](https://stackoverflow.com/questions/54658894/how-do-i-create-a-unique-id-for-a-rust-object) seem to indicate that this is a problematic way to do things.

The better option, I think, is using [`AtomicUsize`](https://doc.rust-lang.org/stable/std/sync/atomic/struct.AtomicUsize.html) or [`AtomicU64`](https://doc.rust-lang.org/stable/std/sync/atomic/struct.AtomicU64.html) as a (semi-)global incrementor for integer ids.

Reply:
Can you expand on how it might be problematic? I can see how the question is similar, but it seems different in that they are using the address of a value, where we are using the address to a value an Arc points to. For an `Arc` to work, I don't see how that could change without every copy of the `Arc` being able to see that change, so I think we are good. The [docs on `ptr_eq`](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.ptr_eq) do callout a gotcha with `dyn Trait`, but that doesn't apply here.